node/pkg/watchers/evm/finalizers/optimism.go

```diff
@@ -0,0 +1,72 @@
+package finalizers
+
+import (
+	"context"
+
+	"github.com/certusone/wormhole/node/pkg/watchers/evm/connectors"
+
+	"go.uber.org/zap"
+)
+
+// OptimismFinalizer implements the finality check for Optimism.
+// Optimism provides a special "rollup_getInfo" API call to determine the latest L2 (Optimism) block to be published on the L1 (Ethereum).
+// This finalizer polls that API to determine if a block is finalized.
+
+type OptimismFinalizer struct {
+	logger    *zap.Logger
+	connector connectors.Connector
+}
+
+func NewOptimismFinalizer(ctx context.Context, logger *zap.Logger, connector connectors.Connector) *OptimismFinalizer {
+	return &OptimismFinalizer{
+		logger:    logger,
+		connector: connector,
+	}
+}
+
+func (f *OptimismFinalizer) IsBlockFinalized(ctx context.Context, block *connectors.NewBlock) (bool, error) {
+	type Result struct {
+		Mode       string
+		EthContext struct {
+			BlockNumber uint64 `json:"blockNumber"`
+			TimeStamp   uint64 `json:"timestamp"`
+		} `json:"ethContext"`
+		RollupContext struct {
+			Index         uint64 `json:"index"`
+			VerifiedIndex uint64 `json:"verifiedIndex"`
+		} `json:"rollupContext"`
+	}
+
+	var m Result
+	err := f.connector.RawCallContext(ctx, &m, "rollup_getInfo")
+	if err != nil {
+		f.logger.Error("failed to get rollup info", zap.String("eth_network", f.connector.NetworkName()), zap.Error(err))
+		return false, err
+	}
+
+	// f.logger.Info("got rollup info", zap.String("eth_network", f.connector.NetworkName()),
+	// 	zap.String("mode", m.Mode),
+	// 	zap.Uint64("l1_blockNumber", m.EthContext.BlockNumber),
+	// 	zap.Uint64("l2_blockNumber", m.RollupContext.Index),
+	// 	zap.Uint64("verified_index", m.RollupContext.VerifiedIndex),
+	// 	zap.Stringer("desired_block", block.Number),
+	// )
+
+	return block.Number.Uint64() <= m.RollupContext.Index, nil
```

Comment:
Placeholder:  I'm not sure this is sufficient.

Reply:
Agreed.  Reading [How Optimism Works](https://community.optimism.io/docs/how-optimism-works/#fault-proofs), it says:

> If a proposed state commitment goes unchallenged for the duration of the challenge window (currently set to 7 days), then it is considered final. Once a commitment is considered final, smart contracts on Ethereum can safely accept proofs about the state of Optimism based on that commitment.

So I think we'd first need to wait for the Eth block to be finalized and then wait another 7 days for the Optimism block to be considered finalized before we can release the VAA.  Since the latency is so bad I assume we'll also need to provide an opt-in mechanism for integrators who want their VAAs faster at the risk of the block being rolled back (basically the equivalent of the consistency level 200 thing we did for Eth).

Reply:
As Evan pointed out: This is true for on-chain anchoring of the L2. Given the optimistic nature the sequencer can submit incorrect new state roots to the root L1 and therefore emit forged messages. This is why there is the dispute period for people to challenge the *execution* results. However the sequence of transactions is finalized (as long as the L1 does not roll back) as that is fully maintained by the L1 SC CanonicalTransactionChain (CTC).
While executing all state transitions (txs) on the L1 would defeat the purpose of the L2, Wormhole guardians can run "full-nodes" that execute the txs in the CTC and are able to derive events and the new state root. Instead of the optimistic model where consensus on the view of the system is achieved by someone proposing the correct result and others raising disputes, wormhole achieves consensus using a MultiSig / majority consensus and achieve "instant finality" on the state root. This means that even if the L2 sequencer misbehaves and an incorrect state root is submitted on-chain, guardian full-nodes that are replaying the CTC would see the correct view to which the on-chain contract would also converge post dispute-resolultion (as long as there is not a bug in the node software).
All that Wormhole needs to check in addition is that the L1 is unlikely to roll back the transaction ordering before a message is confirmed.
The [Fault Detector](https://github.com/smartcontracts/simple-optimism-node/#fault-detector) piece that @evan-gray pointed out might still be helpful as as the L2 enters into dispute resolution mode unexpected issues might occur. This adapter might halt block finalization if the fault detector triggers as a precaution.

Reply:
So your description makes sense to me but the thing I'm having difficulty with is understanding the exact order in which the different components interact with each other in the wormhole context.  I _think_ this is what's happening:

1. The Optimism sequencer publishes blocks to the CTC on Eth.
1. Guardians run full nodes that download the blocks from the sequencer and then replay the execution, performing the relevant state transitions and emitting events.
1. guardiand connects to the guardian full node (via `NewEthereumConnector`) and listens for events from the wormhole contract.
1. Since each guardian trusts their own full node, the VAA has "instant finality" and so can be published immediately.

Is this correct?  At what step do we wait for ETH finality?  Does each full node wait for finality before replaying the L2 block execution (step 2) or is guardiand supposed to wait for ETH finality (step 4)?

Regardless of the answer here I think this interaction is sufficiently complex that it really needs to be documented alongside the actual code.

Reply:
Yes I think your understanding is correct. The wormhole guardian should verify that the ETH block the CTC was submitted on for a certain block is finalized on ETH (which from what I can see is not happening here yet)

Reply:
The ETH block is being checked for finality.